examples/opensora_pku/opensora/acceleration/communications.py

```diff
@@ -0,0 +1,266 @@
+        inp_shape = list(input_.shape)
+        inp_shape[scatter_dim] = inp_shape[scatter_dim] // sp_size
+        if scatter_dim < 1:
+            input_t = input_.reshape([sp_size, inp_shape[scatter_dim]] + inp_shape[scatter_dim + 1 :])
+        else:
+            # transpose groups of heads with the seq-len parallel dimension, so that we can scatter them!
+            input_t = (
+                input_.reshape([-1, sp_size, inp_shape[scatter_dim]] + inp_shape[scatter_dim + 1 :])
+                .swapaxes(0, 1)
+                .contiguous()
+            )
+
+        output = self.alltoall(input_t)
+
+        if scatter_dim < 1:
+            output = output.swapaxes(0, 1).contiguous()
+
+        output = output.reshape(
+            inp_shape[:gather_dim]
+            + [
+                inp_shape[gather_dim] * sp_size,
+            ]
+            + inp_shape[gather_dim + 1 :]
+        )
+
+        return output
+
+
+class bak_SingleAll2ALL(nn.Cell):
+    def __init__(self, scatter_dim: int, gather_dim: int):
+        super(_SingleAll2ALL, self).__init__()
+        self.sp_size = hccl_info.world_size
+        self.spg = hccl_info.group
+        self.scatter_dim = scatter_dim
+        self.gather_dim = gather_dim
+        self.alltoall = ops.AlltoAll(split_count=self.sp_size, split_dim=0, concat_dim=0, group=self.spg)
+
+    def construct(self, input_: Tensor):
+        scatter_dim, gather_dim, sp_size = self.scatter_dim, self.gather_dim, self.sp_size
+        inp_shape = list(input_.shape)
+        inp_shape[scatter_dim] = inp_shape[scatter_dim] // sp_size
+        if scatter_dim < 1:
+            input_t = input_.reshape([sp_size, inp_shape[scatter_dim]] + inp_shape[scatter_dim + 1 :])
+        else:
+            # transpose groups of heads with the seq-len parallel dimension, so that we can scatter them!
+            input_t = (
+                input_.reshape([-1, sp_size, inp_shape[scatter_dim]] + inp_shape[scatter_dim + 1 :])
+                .swapaxes(0, 1)
+                .contiguous()
+            )
+
+        output = self.alltoall(input_t)
+
+        if scatter_dim < 1:
+            output = output.swapaxes(0, 1).contiguous()
+
+        output = output.reshape(
+            inp_shape[:gather_dim]
+            + [
+                inp_shape[gather_dim] * sp_size,
+            ]
+            + inp_shape[gather_dim + 1 :]
+        )
+
+        return output
+
+
+class AllGather(nn.Cell):
+    def __init__(self):
+        super(AllGather, self).__init__()
+        self.op = ops.AllGather(hccl_info.group)
+
+    def construct(self, x):
+        return self.op(x)
+
+
+class AllToAll_SBH(_SingleAll2ALL):
+    def __init__(self, scatter_dim: int = 1, gather_dim: int = 0):
+        super(AllToAll_SBH, self).__init__(scatter_dim=scatter_dim, gather_dim=gather_dim)
+
+
+def prepare_parallel_data(
+    hidden_states, noise, encoder_hidden_states, attention_mask, encoder_attention_mask, use_image_num
+):
```

Comment:
Can you give some docstrings for the function `prepare_parallel_data` ? It alters the inputs' shapes to the LatteT2V and introduces two variables `temp_attention_mask` and `loss_mask`. These are critical changes and some explanations are favorable.

Reply:
done